### PR TITLE
Add lnurl-pay (round 2)

### PR DESCRIPTION
**Pull request by leesalminen**
This PR picks up where https://github.com/GaloyMoney/galoy-mobile/pull/321 left off from @pseudozach. This PR adds support for sending sats with lnurl-pay.

### CODE REVIEW

File: app/screens/send-bitcoin-screen/scanning-qrcode-screen.tsx [hunk 1]
```diff
@@ -81,12 +82,40 @@ export const ScanningQRCodeScreen: ScreenType = ({
...
+          setPending(true)
+          const lnurlParams = await getParams(lnurl)
+
+          if ("reason" in lnurlParams) {
+            throw lnurlParams.reason
+          }
+
+          switch (lnurlParams.tag) {
+            case "withdrawRequest":
+              // TODO: lnurl-withdraw
```
Comment (FuzzWorley) [inline app/screens/send-bitcoin-screen/scanning-qrcode-screen.tsx:97]: thinking about the UX of this, would this cause a flow where the user tries to withdraw and just can't, but there is no message explaining anything? if that's the case, we should just add some kind of message like "lnurl withdrawals not supported yet"

**Comment by daviroo** [comment]
Hi @leesalminen.  Thanks for picking this up.  Unfortunately this isn't working for me when I checked out the code and ran it against production.  I'm attempting to make payments using [ln-url playground](https://lnurl.fiatjaf.com/).  I am getting either a `'toString' is undefined` error when scanning the QR code or I get the following error on the send bitcoin confirmation screen.

image

Were you able to successfully make an ln-url payment using this code?

**Comment by leesalminen** [comment]
Hey @daviroo , my understanding is that ln.fiatjaf.com generates different QR codes on each use, some of which generate error codes, others not. The one with `'toString' is undefined` error I'll need to handle that error more appropriately, but the second one in your screenshot I believe is an expected error code for that LNURL fiatjaf generates.

I've been testing against https://pay.bitcoinjungle.app/lee2/print and https://ln.bitcoinbeach.com/nicolas/print which should never produce an error code and they've worked each time.

I'll continue investigating and post back later today.

**Comment by dolcalmi** [comment]
please don't forget to include `successAction` support https://github.com/fiatjaf/lnurl-rfc/blob/luds/09.md

**Comment by dolcalmi** [comment]
> please don't forget to include `successAction` support https://github.com/fiatjaf/lnurl-rfc/blob/luds/09.md

Reviewing in more depth I think we should include this in an additional PR... if we want to fully comply with rfc then this will need to add an additional mutation to the backend.

**Comment by leesalminen** [comment]
Hi @daviroo, I've just pushed a change that handles the `toString is undefined` error that you encountered. The other error from your screenshot is, AFAICT, an expected error of that specific LNURL-PAY QR code that you scanned from fiatjaf's site. Every time I refresh that page, I get a new QR code that changes up various parameters - min/max sats, comment required, etc. Some of them are intended to fail, for example if the `payment secret` is missing. So, you'll want to try out various codes generated by fiatjaf's page to test various scenarios, as well as using ln.bitcoinbeach.com / pay.bitcoinjungle.app to get always valid LNURL codes.

File: app/screens/send-bitcoin-screen/send-bitcoin-screen.tsx [hunk 2]
```diff
@@ -257,11 +293,37 @@ export const SendBitcoinScreen: ScreenType = ({
     setDestinationStatus("NOT_CHECKED")
   }, [destination])
 
-  const pay = useCallback(() => {
+  const fetchInvoice = async (url) => {
+    const response = await fetch(url)
+    return await response.json()
```
Comment (dolcalmi) [inline app/screens/send-bitcoin-screen/send-bitcoin-screen.tsx:298]: not sure about the internals of `fetch` but if it throws then the app will crash. We can catch the errors here an return proper `status` and `reason`

File: app/screens/send-bitcoin-screen/scanning-qrcode-screen.tsx (hunk 1 above)
Comment (leesalminen) [reply app/screens/send-bitcoin-screen/scanning-qrcode-screen.tsx]: Great feedback, thanks @FuzzWorley, I've added user friendly error messaging here.

File: app/screens/send-bitcoin-screen/send-bitcoin-screen.tsx (hunk 2 above)
Comment (leesalminen) [reply app/screens/send-bitcoin-screen/send-bitcoin-screen.tsx]: Good catch (lol), @dolcalmi. I've just pushed up a fix for this.

File: app/screens/send-bitcoin-screen/send-bitcoin-screen.tsx [hunk 3]
```diff
@@ -257,11 +293,50 @@ export const SendBitcoinScreen: ScreenType = ({
...
+
+  const pay = useCallback(async () => {
     if (paymentType === "username" && destinationStatus !== "VALID") {
       userDefaultWalletIdQuery({ variables: { username: destination } })
       toastShow(translate("SendBitcoinScreen.usernameNotFound"))
       return
+    } else if (paymentType == "lnurl") {
+      const satAmount = primaryAmount.value * 1000
+      const lnurlInvoice = await fetchInvoice(
+        `${lnurlPay.callback}?amount=${satAmount}&comment=${memo}`,
```
[the remark this answers is not in the document]
Comment (leesalminen) [reply app/screens/send-bitcoin-screen/send-bitcoin-screen.tsx:322]: Hey @daviroo, thanks for pointing this out. My understanding was that `fetch()` will urlencode the URL without any specific intervention. I did an test from my browser where I ran `fetch('https://leesalminen.com?memo=this is a test memo. it should be url-encoded. has spaces, special characters ***, hola, mañana')` and the actual HTTP request that was sent was `https://leesalminen.com/?memo=this%20is%20a%20test%20memo.%20it%20should%20be%20url-encoded.%20has%20spaces,%20special%20characters%20***,%20hola,%20ma%C3%B1ana`. See screenshot attached.

Did you run into any error adding a memo with special characters in-app?

Screen Shot 2022-01-04 at 10 48 15 AM

[the remark this answers is not in the document]
Comment (leesalminen) [reply app/screens/send-bitcoin-screen/send-bitcoin-screen.tsx:322]: Also ran a test from `node-fetch` on `v.2.6.0`. 

```
import fetch from 'node-fetch';

fetch('https://leesalminen.com?memo=this is a test memo. it should be url-encoded. has spaces, special characters ***, hola, mañana')
.then((res) => {
	console.log(res)
})
.catch((err) => {
	console.log(err)
})
```

the console shows:

```
Response {
  size: 0,
  timeout: 0,
  [Symbol(Body internals)]: {
    body: PassThrough {
      _readableState: [ReadableState],
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      _writableState: [WritableState],
      allowHalfOpen: true,
      [Symbol(kCapture)]: false,
      [Symbol(kCallback)]: null
    },
    disturbed: false,
    error: null
  },
  [Symbol(Response internals)]: {
    url: 'https://leesalminen.com/?memo=this%20is%20a%20test%20memo.%20it%20should%20be%20url-encoded.%20has%20spaces,%20special%20characters%20***,%20hola,%20mañana',
    status: 400,
    statusText: 'Bad Request',
    headers: Headers { [Symbol(map)]: [Object: null prototype] },
    counter: 0
  }
}
```

[the remark this answers is not in the document]
Comment (leesalminen) [reply app/screens/send-bitcoin-screen/send-bitcoin-screen.tsx]: Hey @daviroo, ignore the last comment, you were correct. There seems to be a difference between the browser's implementation of `fetch` and node's. I've pushed up a change that wraps the `memo` in `encodeURIComponent()`.

[the remark this answers is not in the document]
Comment (daviroo) [reply app/screens/send-bitcoin-screen/send-bitcoin-screen.tsx]: Hey @leesalminen.  Yeah I thought the same regarding the behaviour of fetch but then I googled it and saw the difference between browser and node.  Right now the memo isn't working for the lightning invoice but I think that's something internal for us to fix.

**Review by daviroo — APPROVED** [review]

**Pull request merged**